File: DFS:BFS/bfs.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

void bfs(int start, vector<int> v[], bool visited[]){
    queue<int> q;
    q.push(start);
    visited[start] = true;

    while(!q.empty()){
        int node = q.front();
        q.pop();
        cout << node << endl;

        for (int i = 0; i < v[node].size(); i++){
            int next = v[node][i];

            if (visited[next] == false){
                q.push(next);
                visited[next] = true;
            }
    }
}

int main(void){
    vector<int> a[10];
    bool visited[9];

    a[1].push_back(2);
    a[2].push_back(1);

    a[1].push_back(3);
    a[3].push_back(1);

    a[2].push_back(3);
    a[3].push_back(2);

    a[2].push_back(4);
    a[4].push_back(2);

    a[2].push_back(5);
    a[5].push_back(2);

    fill(visited, visited+10, false);

    bfs(1, a, visited);

    return 0;
}

File: DFS:BFS/bfs.py
# 너비 우선 탐색 
# 주로 두 노드 사이에서 최단 경로 알고 싶을 때


def bfs(graph, node):
    need_visited, visited = list(), list()

    need_visited.append(node)
    
    while need_visited:
        node = need_visited[0]
        del need_visited[0]

        if node not in visited:
            visited.append(node)
            need_visited.extend(graph[node])
    
    return visited


if __name__ == "__main__":
    graph = dict()
 
    graph['A'] = ['B', 'C']
    graph['B'] = ['A', 'D']
    graph['C'] = ['A', 'G', 'H', 'I']
    graph['D'] = ['B', 'E', 'F']
    graph['E'] = ['D']
    graph['F'] = ['D']
    graph['G'] = ['C']
    graph['H'] = ['C']
    graph['I'] = ['C', 'J']
    graph['J'] = ['I']

    print(bfs(graph, 'A'))